dbfopen.c

```diff
@@ -1744,7 +1744,7 @@ int SHPAPI_CALL DBFIsRecordDeleted( DBFHandle psDBF, int iShape )
 /* -------------------------------------------------------------------- */
 /*      Verify selection.                                               */
 /* -------------------------------------------------------------------- */
-    if( iShape < 0 || iShape >= psDBF->nRecords )
+    if( iShape < 0 || (iShape > 0 && iShape >= psDBF->nRecords) )
```

Comment:
Shouldn't that be ? (I changed the iShape > 0 to psDBF->nRecords > 0 since that matches more your rationale + added the comment)
```suggestion
    /* Some DBF files are written without a record count (because, say */
    /* the SHP file or SHX file already give a record count. In that */
    /* case, checking the shape index being > number of records always */
    /* returns deleted, which is not correct. */
    if( iShape < 0 || ( psDBF->nRecords > 0 && iShape >= psDBF->nRecords) )
```
I didn't check the DBF spec but I don't think a DBF header with nRecords == 0 would be valid ?
And as far as I can see DBFReadAttribute() would have the same issue.,  and quite a few other places How much is your proposed change tested ;-) ?

Reply:
Ha ha. Well, in theory it's been tested widely in PostGIS forever, but I only had to make the once change to pass regression... and your change is correct, the comment is write and my code is not. I think our test files are so simple we aren't even seeing the DBFReadAttribute case because we have no attributes in our DBF files.

Reply:
Is there a point in trying to support such broken DBF files, unless you see them in great use in the wild ? I don't have memories of issues reported in GDAL regarding such cases

Reply:
It's possible we are generating them ourselves... leave this PR open, I will check tomorrow... somehow we got these shape files in the very early days and added them to our regression suite... https://github.com/postgis/postgis/tree/master/regress/loader

Reply:
Either we generated them ourselves with early versions of shapelib or maybe created them with ArcView 3.

Reply:
maybe regenerate them with a recent ogr2ogr ?